### ISSUE TITLE
[Bug] Remote certificate is invalid according to the validation procedure when using Microsoft.Data.SqlClient 2.0.1

### ISSUE THREAD

**Issue by liquidcowgithub** [AzureAD/microsoft-authentication-library-for-dotnet#2054]
**Which Version of MSAL are you using ?**
Microsoft.Identity.Client (>= 4.14.0) as a dependency of Microsoft.Data.Sqlclient 2.0.1

**Platform**
netcore - .net core 3.1.8 on mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic images on AKS, Microsoft.Data.SqlClient 2.0.1

**What authentication flow has the issue?**
* Desktop / Mobile
    * [ ] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App
    * [ ] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

Other? - please describe;
Azure SQLDB authentication using a service principal via Microsoft.Data.Sqlclient 2.0.1
https://github.com/dotnet/SqlClient/blob/master/release-notes/2.0/2.0.0.md#active-directory-service-principal

**Is this a new or existing app?**
c. This is a new app or experiment

**Repro**

New .net core app, with ef core to Azure SQLDB instance, using the following connection string format:

```
Server=tcp:<servername>.database.windows.net;Database=<dbname>;Authentication=Active Directory Service Principal;User Id=<AppId>;Password=<Secret>;TrustServerCertificate=True;
```

Set a self-signed certificate via environment variables:
ASPNETCORE_Kestrel__Certificates__Default__Path
ASPNETCORE_Kestrel__Certificates__Default__Password

Original issue: https://github.com/dotnet/SqlClient/issues/738

**Expected behavior**
Connection to Azure SQLDB instance is established

**Actual behavior**
An error occurs:

```
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.
```

Full stack trace below.

The error is correct, in that a corporate issued self-signed certificate is in use, however, this is not the behaviour that's observed when SQL authentication is used, with the same self-signed certificate and the "TrustServerCertificate=True" property in the connection string. With SQL authentication it connects fine.
Additionally, running ef core migrations with "Authentication=Active Directory Service Principal;" works fine (via Azure DevOps)

**Possible Solution**

- Use SQL authentication instead of Active Directory Service Principal - Not desirable
- Don't use a self-signed certificate - Desirable, but contextually not an option (and is in effect the root cause of the issue)
- Figure out how to supply a customised HttpClient to Microsoft.Data.Sqlclient (as per [docs](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/httpclient))
- Possibly [register the certificate with Microsoft identity platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-certificate-credentials#register-your-certificate-with-microsoft-identity-platform) - undocumented, uncertain if this is a viable solution, needs some experimentation?

**Additional context/ Logs / Screenshots**
[stack trace.txt](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/files/5274202/stack.trace.txt)

Scenario is a corporate with a self-signed certificate issued to a web app, that needs to authenticate with an Azure SQLDB instance using an AD service principal.

This is probably not a bug, but rather a use case that's common in some corporates in the process of migrating to the cloud or in a hybrid of on-prem and cloud, and there's no obvious way to alter the configuration of token acquisition to support this? Unless my google-fu is failing me?

Conscious that self-signed and invalid certificate chains are just generally not a good idea. So maybe the right thing to do here is to just explicitly not support this use case, and indirectly dissuade usage of invalid certificates?

Original issue reference from Microsoft.Data.Sqlclient: https://github.com/dotnet/SqlClient/issues/738

**Comment by jmprieur** [comment]
@liquidcowgithub 
Thanks for raising this issue. I understand the repro, however, I don't understand how this is actionable for the MSAL.NET team? I don't see either that MSAL.NET has a dependency on `Microsoft.Data.Sqlclient 2.0.1`. (It's the contrary actually).

![image](https://user-images.githubusercontent.com/13203188/94121781-0a315500-fe52-11ea-9c48-ed9e47ac7fcc.png)

I would think that you'd want to raise this issue against Microsoft.Data.SqlClient? not MSAL.NET ?
MSAL.NET provides a mechanism to provide your own HttpClientFactory. 
See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/httpclient
Would it be useful for Microsoft.Data.SqlClient to use it?

Unless, I miss something, I'm proposing to classify this bug as external. But feel free to provide more details.

**Comment by liquidcowgithub** [comment]
Thanks @jmprieur,

I reached out here on suggestion from discussion on the issue I raised against Microsoft.Data.SqlClient ([738](https://github.com/dotnet/SqlClient/issues/738)) primarily just to air the use case, and to check that the observed behaviour is expected.

On looking through the docs when opening this issue, I did see the httpclient mechanism,  https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/httpclient, and am exploring this, will revert if I find more information. 

I agree that the issue possibly falls outside the scope of both packages, and is more of an integration issue between the two?

Let me see if I can get more information on what Microsoft.Data.SqlClient's AD Service Principal dependency tree looks like.
